File: daily/cpp/2485_find_the_pivot_integer.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
 public:
  int pivotInteger(int n) {
    // 1. use the math directly, sqrt((n * n + n) / 2)
    int ans = sqrt((n * n + n) / 2);
    if (ans * ans == ((n * n + n) / 2)) {
      return ans;
    }
    return -1;

    // 2. traverse from 1 to n, use vector to store
    if (n == 1) return 1;
    vector<int> nums;
    for (int i = 0; i < n; ++i) {
      nums.push_back(i + 1);
    }
    int sum = 0;
    vector<int> sums;
    for (int i = 0; i < n; ++i) {
      sum += nums[i];
      sums.push_back(sum);
    }

    int totalSum = 0;
    totalSum = sum;
    sum = 0;
    for (int i = 0; i < n; ++i) {
      if (sum == totalSum - sum - nums[i]) {
        return i + 1;
      }
      sum += nums[i];
    }
    return -1;
  }
};

File: daily/rust/2485_find_the_pivot_integer.rs
impl Solution {
    pub fn find_pivot(n: i32) -> i32 {
        // math
        // use sqrt to find the pivot
        let mut ans = 0;
        ans = (((n * n + n) / 2) as f64).sqrt() as i32;
        if ans * ans == (n * n + n) / 2 {
            return ans;
        }
        -1

        // traverse
        let mut sum_forward: i32 = (1 + n) * n / 2;
        let mut sum_backward: i32 = 0;
        let mut n: i32 = n;
        let mut ans: i32 = -1;

        while (n > 0) {
            sum_backward += n;

            if sum_backward == sum_forward {
                ans = n;
                break;
            }

            sum_forward -= n;
            n -= 1;
        }
        ans
    }
}